### PR TITLE
ci: migrating from CircleCI 1.0 to 2.0

### PR DESCRIPTION
**Pull request by Librazy**
For #2979
Introduced a basic CircleCI  2.0 config with a single job. The workflow is consistent with previous one.

This PR utilized built-in CircleCI cache by caching all things in ~/.cargo ./target etc.

TODO:
- [ ] Optimize CI jobs and workflow
- [ ] Investigate whether we can use a docker image instead of setup build environment manually
- [ ] Investigate and setup sccache for building and testing if needed

### CODE REVIEW

**Comment by siddontang** [comment]
Cool @Librazy, thank you.

**Comment by siddontang** [comment]
PTAL @BusyJay @Hoverbear

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            - v1-local-{{ .Branch }}
+            - v1-local
+      - restore_cache:
+          keys:
+            - v1-cargo-{{ .Branch }}-{{ checksum "Cargo.lock" }}
+            - v1-cargo-{{ .Branch }}
+            - v1-cargo
+
+      - run: 
+          name: Setting up Rust and Other Build Environment
```
Comment (Hoverbear) [inline .circleci/config.yml]: "Setting up the build environment" or "Setting up Rust and the rest of the build environment"

**Comment by Hoverbear** [comment]
Why do we see

>  ci/circleci Expected — Waiting for status to be reported 

For this PR?

File: .circleci/config.yml (hunk 1 above)
Comment (Librazy) [reply .circleci/config.yml]: Thank you advice!

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      - restore_cache:
+          keys:
+            - v1-target-{{ .Branch }}-{{ .Environment.CIRCLE_SHA1 }}
+            - v1-target-{{ .Branch }}
+            - v1-target
+      - run: 
+          name: Building
+          command: env SKIP_TESTS=true make trace_test
+          no_output_timeout: 1800s
+      - save_cache:
```
Comment (Hoverbear) [inline .circleci/config.yml]: Can you please give these names so they do not show up as the same name in the CI log?

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            echo 'export LD_RUN_PATH=$LD_RUN_PATH:$HOME/.local/lib' >> $BASH_ENV
+            source $BASH_ENV
+      - checkout
+      - run: 
+          name: Installing Build Dependencies
+          command: |
+            sudo add-apt-repository ppa:ubuntu-toolchain-r/test -y;
+            sudo apt-get update -y;
+            sudo apt-get install -y g++-4.8 curl libdw-dev libelf-dev elfutils binutils-dev
+      - restore_cache:
```
Comment (Hoverbear) [inline .circleci/config.yml]: Can you please give these names so they do not show up as the same name in the CI log?

File: .circleci/config.yml (hunk 2 above)
Comment (Librazy) [reply .circleci/config.yml]: I try to do so but it seems no `name` available for `save_cache` https://circleci.com/docs/2.0/configuration-reference/#save_cache

**Comment by Librazy** [comment]
For `ci/circleci Expected`, `raftstore_cases::test_conf_change::test_node_learner_conf_change` spurious failed at the first run for this PR (`https://circleci.com/gh/Librazy/tikv/13`). I restarted the CI build and everything returned normal but the `ci/circleci` remains `Expected — Waiting for status to be reported` from then on. IMO it is a bug of Github's webhook and doesn't matter.

**Comment by Hoverbear** [comment]
Okay. Thank you for looking into this. :heart_eyes:

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+              cd gflags-2.1.2 && \
+              cmake -DCMAKE_INSTALL_PREFIX=${LOCAL_PREFIX} . && \
+              make -j && \
+              make install;
+            fi
+      # check format first
+      -  run:
+          name: Checking Format
+          command: make format && git diff-index --quiet HEAD -- || (git diff; echo please make format and run tests before creating a PR!; exit 1)
+      - restore_cache:
```
Comment (Hoverbear) [inline .circleci/config.yml]: Does it not work to restore the `target` before running the format?

Comment (Librazy) [reply .circleci/config.yml]: Yes, it works but restoring the `./target` and maybe `sccache` for the future may take a long time to process (about five minutes), checking format first may reduce feedback time for the PR authors.

Comment (Hoverbear) [reply .circleci/config.yml]: That's a great reason. :+1:

**Comment by Hoverbear** [comment]
@Librazy I noticed just now in the `Testing` step on the CI there is a `/bin/sh` error while running `make trace_test`. Can you please take a look at it?

Right at the bottom:

```
/bin/sh: 6: [[: not found
# TODO: remove above target once https://github.com/rust-lang/cargo/issues/2984 is resolved.
make[1]: Leaving directory `/home/circleci/tikv-ci'
```

**Comment by Librazy** [comment]
There is same issue in CircleCI 1, and I think it is introduced in #2721 . Maybe we need to prepend `SHELL=/bin/bash -o pipefail` to https://github.com/pingcap/tikv/blob/master/Makefile ?

**Comment by Hoverbear** [comment]
Great job, thanks for this!

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      RUST_TEST_THREADS: "1"
+      RUST_BACKTRACE: "1"
+      RUSTFLAGS: "-Dwarnings"
+      FAIL_POINT: "1"
+    steps:
+      - run:
+          name: Updating PATH and Define Environment Variable at Runtime
+          # Define variables that need interpolation
+          command: |
+            echo 'export PATH=$HOME/.local/cmake/bin:$HOME/.cargo/bin:$HOME/.local/bin:$PATH' >> $BASH_ENV
```
Comment (BusyJay) [inline .circleci/config.yml]: Why not just `export`?

File: Makefile [hunk 6]
```diff
@@ -59,6 +59,7 @@ fail_release:
 trace_test:
 	env CI=true SKIP_FORMAT_CHECK=true FAIL_POINT=1 ${PROJECT_DIR}/ci-build/test.sh
 
+test: SHELL=/bin/bash -o pipefail
```
Comment (BusyJay) [inline Makefile]: Why? I don't see any pipes in this target.

Comment (Librazy) [reply Makefile]: Good point. Will fix it.

**Comment by Librazy** [comment]
There are many builds that failed with `raftstore::store::util::tests::test_lease`. Am I missing anything?

File: .circleci/config.yml (hunk 5 above)
Comment (BusyJay) [reply .circleci/config.yml]: Then why source it at L27?

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            echo 'export LD_RUN_PATH=$LD_RUN_PATH:$HOME/.local/lib' >> $BASH_ENV
+            echo 'export DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:$HOME/.local/lib' >> $BASH_ENV
+            echo 'export LDFLAGS=-L$HOME/.local/lib' >> $BASH_ENV
+            echo 'export CPPFLAGS=-I$HOME/.local/include' >> $BASH_ENV
+            echo 'export CXXFLAGS=-I$HOME/.local/include' >> $BASH_ENV
+            echo 'export LOCAL_PREFIX=$HOME/.local' >> $BASH_ENV
+            echo 'export LD_RUN_PATH=$LD_RUN_PATH:$HOME/.local/lib' >> $BASH_ENV
```
Comment (BusyJay) [inline .circleci/config.yml]: Duplicated lines with L19 and L26.

File: .circleci/config.yml [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+            make -f ci-build/Makefile prepare-rustfmt
+
+            if [[ ! -f $HOME/.local/cmake/bin/cmake ]]; then
+              curl https://cmake.org/files/v3.10/cmake-3.10.0-Linux-x86_64.tar.gz -sSf | tar xzf -;
+              rm -rf $HOME/.local/cmake
+              mv cmake-3.10.0-Linux-x86_64 $HOME/.local/cmake
+            fi
+
+            if [[ ! -e $HOME/.local/lib/libgflags.a ]]; then
```
Comment (BusyJay) [inline .circleci/config.yml]: I don't think it's necessary anymore.

File: .circleci/config.yml (hunk 7 above)
Comment (Librazy) [reply .circleci/config.yml]: My bad, fixed.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**